akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -242,27 +242,47 @@ private[stream] object Collect {
   override def initialAttributes: Attributes = DefaultAttributes.collect and SourceLocation.forLambda(pf)
 
   def createLogic(inheritedAttributes: Attributes) =
-    new SupervisedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
-
-      import Collect.NotApplied
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private var buffer: OptionVal[Out] = OptionVal.none
 
-      val wrappedPf = () => pf.applyOrElse(grab(in), NotApplied)
+      override def preStart(): Unit = pull(in)
 
-      override def onPush(): Unit = withSupervision(wrappedPf) match {
-        case Some(result) =>
-          result match {
-            case NotApplied             => pull(in)
-            case result: Out @unchecked => push(out, result)
-            case _                      => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
+      def onPush(): Unit =
+        try {
+          val elem = grab(in)
+          import Collect.NotApplied
+          pf.applyOrElse(elem, NotApplied) match {
+            case NotApplied => pull(in)
+            case result: Out @unchecked =>
+              if (isAvailable(out)) {
+                push(out, result)
+                pull(in)
+              } else {
+                buffer = OptionVal.Some(result)
+              }
+            case _ => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
           }
-        case None => //do nothing
-      }
+        } catch {
+          case NonFatal(ex) =>
+            decider(ex) match {
+              case Supervision.Stop => failStage(ex)
+              case _                => if (!isClosed(in)) pull(in) else completeStage()
+            }
+        }
 
-      override def onResume(t: Throwable): Unit = if (!hasBeenPulled(in)) pull(in)
+      override def onPull(): Unit = buffer match {
+        case OptionVal.Some(value) =>
+          push(out, value)
+          buffer = OptionVal.none
+          if (!isClosed(in)) pull(in) else completeStage()
+        case _ => // already pulled
+      }
 
-      override def onPull(): Unit = pull(in)
+      override def onUpstreamFinish(): Unit = if (buffer.isEmpty) super.onUpstreamFinish()
 
       setHandlers(in, out, this)
+
```

Comment:
It's a somewhat scary amount of changes (esp the early pull) vs just discard-pull when filtered out, but I see that it aligns with what filter does so maybe that is fine.

Reply:
Could there event be an advantage of having `collect` as it was? This new behaviour could be achieved with `filter+map`?

We also have dropWhile (and maybe more) that would be similar?

Reply:
![image](https://user-images.githubusercontent.com/501740/192248419-26ce827b-6cfe-4d0b-b02f-883ea7acb18f.png)